### PR TITLE
Starting the interpretability of the Deep Forest using SHAP

### PR DESCRIPTION
**Pull request by Maryom**
Hey,

This is an initial implementation, however I'm not sure it will work I see that we will get the following error:

> AttributeError: 'CascadeForestClassifier' object has no attribute 'estimators_'

What do you think @xuyxu ?

### CODE REVIEW

**Comment by xuyxu** [comment]
Thanks @Maryom! Never mind, I will switch into your PR and take a careful look ;-)

File: deepforest/_explainer.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        The forest estimator that we want to explain.
+
+    Returns
+    -------
+    TreeExplainer : :obj:`shap.TreeExplainer`
+        Tree explainer for the forest estimator.
+    """
+
+    # Pull the info of the first tree
+    tree_tmp = forest.estimators_[0][0].tree_
```
Comment (xuyxu) [inline deepforest/_explainer.py:32]: I think the problem is here. The forest structure in DF21 does not use a list `estimators_` to store all decision trees.

File: deepforest/_explainer.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    # extract the arrays that define the tree
+    children_left1      = tree_tmp.children_left
```
Comment (xuyxu) [inline deepforest/_explainer.py:35]: According to the forest structure in DF21, this code snippet should be:

```python
children_left1 = forest.childrens[0][:, 0]
children_right1 = forest.childrens[0][:, 1]
features1 = forest.features[0]
thresholds1 = forest.thresholds[0]
values1 = forest.values[0]
```

In additional, forest in DF21 does not record the property `weighted_n_node_samples`

Comment (Maryom) [reply deepforest/_explainer.py]: Great 🙏🏼 hmmm what should we do with `weighted_n_node_samples`?

**Comment by xuyxu** [comment]
Hi @Maryom, could you add the script you are using to test this function? You can put it into another file named `test_explainer.py` in the folder `tests`.

**Comment by Maryom** [comment]
Hi,

Currently there is an error with DF I can't fit it I got the following error:
```
Traceback (most recent call last):
  File "test_explainer.py", line 14, in <module>
    model.fit(X_train, y_train)
  File "/deepforest/cascade.py", line 951, in fit
    super().fit(X, y, sample_weight)
  File "/deepforest/cascade.py", line 484, in fit
    X_train_ = self._bin_data(binner_, X, is_training_data=True)
  File "/deepforest/cascade.py", line 405, in _bin_data
    X_binned = binner.fit_transform(X)
  File "/Library/Python/3.8/lib/python/site-packages/sklearn/base.py", line 699, in fit_transform
    return self.fit(X, **fit_params).transform(X)
  File "/deepforest/_binner.py", line 130, in fit
    self.bin_thresholds_ = _find_binning_thresholds(
  File "/deepforest/_binner.py", line 82, in _find_binning_thresholds
    X[:, f_idx], n_bins, bin_type
  File "/Library/Python/3.8/lib/python/site-packages/pandas/core/frame.py", line 3024, in __getitem__
    indexer = self.columns.get_loc(key)
  File "/Library/Python/3.8/lib/python/site-packages/pandas/core/indexes/base.py", line 3080, in get_loc
    return self._engine.get_loc(casted_key)
  File "pandas/_libs/index.pyx", line 70, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/index.pyx", line 75, in pandas._libs.index.IndexEngine.get_loc
TypeError: '(slice(None, None, None), 0)' is an invalid key
```

**Comment by xuyxu** [comment]
Thanks for reporting ;-). Does shap always use pandas dataframe as the input data? DF21 works the best when the input data is a Numpy array...

**Comment by Maryom** [comment]
Oops sorry I forgot to convert it. Now I got this error:

> Exception: Model type not yet supported by TreeExplainer: <class 'deepforest.cascade.CascadeForestClassifier'>

**Comment by xuyxu** [comment]
We should use the following code snippt ;-)

``` python
import shap
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_iris

from deepforest import CascadeForestClassifier
from deepforest._explainer import get_shap_explainer_forest


# load JS visualization code to notebook
shap.initjs()

X, y = load_iris(return_X_y=True)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.42)

model = CascadeForestClassifier()

model.fit(X_train, y_train)

forest = model.get_forest(0, 0, "rf")

explainer = get_shap_explainer_forest(forest)
```

The reported error is `KeyError: 'node_sample_weight'`. I will take a creaful look at the traceback information, and reply to you ASAP.

**Comment by Maryom** [comment]
Great THANK you @xuyxu I appreciate your help 🙏🏼

**Comment by Maryom** [comment]
I have a question if you don't mind: what is the size of the sliding window in multi-grained scanning in DF21? it's 100?

**Comment by xuyxu** [comment]
After adding the sklearn backend for DF21 (#36), I think we can move on ;-). The following code snippet should work well now:

``` python
import shap
import numpy as np
from deepforest import CascadeForestClassifier
from sklearn.model_selection import train_test_split

# load JS visualization code to notebook
shap.initjs()

X, y = shap.datasets.iris()

X = np.array(X)
y = np.array(y)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

model = CascadeForestClassifier(backend="sklearn")
model.fit(X_train, y_train)
forest = model.get_forest(0, 0, "rf")
explainer = shap.TreeExplainer(forest)
shap_values = explainer.shap_values(X_test)
shap.summary_plot(shap_values, X_test)
```

The running result looks like this:

screenshot


Since I am new to SHAP, would you tell me that what kind of information can we get from this image? Thanks!

**Comment by Maryom** [comment]
Hi @xuyxu thank you so much this is really helpful I appreciate it.

> Since I am new to SHAP, would you tell me that what kind of information can we get from this image? Thanks!

Kindly check [this](https://stackoverflow.com/questions/65549588/shap-treeexplainer-for-randomforest-multiclass-what-is-shap-valuesi)

Currently the SHAP figure shows the four classifiers as features I think we have to fix it. How we can access MGS or the input feature vector?

**Comment by xuyxu** [comment]
Thanks for your kind explainations! So, all we need to do is providing users with a `Explainer`, right?

It is up to the users to conduct post-analysis on the model via calling different methods of this `Explainer`.

**Comment by Maryom** [comment]
> Thanks for your kind explanations! So, all we need to do is providing users with an `Explainer`, right?

I have to thank you for your time and efforts.
Yes, you are right.

> It is up to the users to conduct post-analysis on the model via calling different methods of this `Explainer`.

Yup.

Thanks again 🙏🏼

**Comment by xuyxu** [comment]
Great, I will do a little planing, and we can formally work on this feature request then ;-)

**Comment by Maryom** [comment]
@xuyxu This is awesome

**Comment by Lizheng068** [comment]
Hi. I did not find the code of multi-grained scanning scanning part in DF21, and did not find sliding slicing of input data in the my test? Does the function of multi-grained scanning scanning part exist in DF21? If so, where can I find the corresponding file? Look forward to your help. Thank you!

**Comment by xuyxu** [comment]
Hi @Maryom, I am going to close this PR for now, we may need more verifications before officially working on this feature. Anyway, thanks for your advice on the interpretability and using the customized estimator 😄.

**Comment by Maryom** [comment]
Hi @xuyxu,
I have one question if you don't mind:
Why `get_estimator` is NOT working when model `set_estimator `?

I tried:
```
forest = model.get_estimator(0, 0, "rf")
```
I got: 
```
KeyError: '0-0-rf'
```
I think the issue is with layer because I check the layer, it is when `set_estimator ` is used:

```
<deepforest._layer.CustomCascadeLayer object at 0x1328ccc40>
```
And it is when `set_estimator ` is NOT used:

```
ClassificationCascadeLayer(backend='sklearn',
                           buffer=<deepforest._io.Buffer object at 0x103f98730>,
                           criterion='gini', layer_idx=0, n_estimators=4,
                           n_outputs=3)
```

**Comment by xuyxu** [comment]
Hi @Maryom, the third parameter for `get_estimator` should be `custom` when `set_estimator` was called, you can check the docstrings [here](https://github.com/LAMDA-NJU/Deep-Forest/blob/master/deepforest/cascade.py#L1142).

**Comment by Maryom** [comment]
Hi @xuyxu I got this error:

```
if estimator_type == "custom" and not self.use_custom_estimator:
AttributeError: 'CascadeForestClassifier' object has no attribute 'use_custom_estimator' 
```

**Comment by xuyxu** [comment]
> Hi @xuyxu I got this error:
> 
> ```
> if estimator_type == "custom" and not self.use_custom_estimator:
> AttributeError: 'CascadeForestClassifier' object has no attribute 'use_custom_estimator' 
> ```

Could you post your script on using the customized estimator feature, thanks!

**Comment by Maryom** [comment]
Yes sure:

```
import shap
import numpy as np
from xgboost import XGBClassifier
from sklearn.ensemble import RandomForestClassifier
from deepforest import CascadeForestClassifier
from sklearn.model_selection import train_test_split

# load JS visualization code to notebook
shap.initjs()

X, y = shap.datasets.iris()

X = np.array(X)
y = np.array(y)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

model = CascadeForestClassifier(backend="sklearn")

estimators = [RandomForestClassifier(), RandomForestClassifier(), XGBClassifier(), XGBClassifier()]  # base estimators per cascade layer

model.set_estimator(estimators)

model.fit(X_train, y_train)


forest = model.get_estimator(0, 0, "custom")

explainer = shap.TreeExplainer(forest)
shap_values = explainer.shap_values(X_test)
shap.summary_plot(shap_values, X_test)

```

**Comment by xuyxu** [comment]
Hi @Maryom, sorry for the late response due to the holiday.

I have tested your script (with last three lines commented out), and no error occurs. The variable `forest` is a list of `RandomForestClassifier`. What's the version of deep forest in your environment?

**Comment by Maryom** [comment]
Hey @xuyxu thanks for your reply, it's totally fine TYT.

`(with last three lines commented out)`
I need the last three lines specifically.

`What's the version of deep forest in your environment?`
The latest version.

**Comment by xuyxu** [comment]
> Hey @xuyxu thanks for your reply, it's totally fine TYT.
> 
> `(with last three lines commented out)`
> I need the last three lines specifically.
> 
> `What's the version of deep forest in your environment?`
> The latest version.

I think you can pass each element of `forest` into the SHAP explainer, such as `explainer = shap.TreeExplainer(forest[0])`, since the only difference between them is that they have different training sets (caused by cross-validation). Nevertheless, the central tendency on important features should be similar. Feel free to share the results if you are going to try this ;-)

**Comment by Maryom** [comment]
Great, it worked 😊 Thank you 🙏🏼

Here is the code:

```
forest = model.get_estimator(0, 0, "custom")

explainer = shap.TreeExplainer(forest[0])
shap_values = explainer.shap_values(X_test)
shap.summary_plot(shap_values[0], X_test, feature_names=feature_names)
```

**Pull request closed without merging**